mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp

```diff
@@ -1237,6 +1237,67 @@ struct GridwiseAttentionAccelRewritePattern
         });
   }
 
+  // If padding is used in the kernel, this means the first gemm
+  // will be done in a larger matrix. In typical, gemm kernels
+  // the padded region in the output will just contain zeros. However,
+  // attention kernel will perform softmax normalization on rows.
+  // Therefore, having zeros -- zero not being the minimum representable
+  // value in the element type -- going to affect all the values
+  // post normalization. Therefore, this function creates a trasnforming
+  // for loop that overwrites out of bounds values of first gemm output
+  // to be negative infinity.
+  void createFirstGemmNegInfPadding(PatternRewriter &rewriter, Location loc,
+                                    layout::GridCoordinates gridCoords,
+                                    Value gemm0OutBuffer,
+                                    RegsAsMatrixSubTiles gemm0OutSubTileViews,
+                                    int64_t prePadGemmM,
+                                    int64_t prePadGemmN) const {
+    Value zero = rewriter.createOrFold<ConstantIndexOp>(loc, 0);
+    Value prePadGemmMVal =
+        rewriter.createOrFold<ConstantIndexOp>(loc, prePadGemmM);
+    Value prePadGemmNVal =
+        rewriter.createOrFold<ConstantIndexOp>(loc, prePadGemmN);
+    MemRefType gemm0OutBufferType = gemm0OutBuffer.getType().cast<MemRefType>();
+    auto negInfTyped = createConstantFloatOp(
+        rewriter, loc, gemm0OutBufferType.getElementType(),
+        gemm0OutBufferType.getElementType(),
+        -std::numeric_limits<float>::infinity());
+    // Get current workitem ID.
+    auto tid = rewriter.create<WorkitemIdOp>(loc, rewriter.getIndexType());
+    int64_t elementsInThreadBuffer = gemm0OutBufferType.getNumElements();
+
+    auto loop = rewriter.create<TransformingForOp>(
+        loc,
+        ArrayRef<ValueRange>{{gridCoords.g_block, gridCoords.m_block,
+                              gridCoords.n_block, tid, zero},
+                             {zero, zero, zero, zero, zero}},
+        ArrayRef<Attribute>{gemm0OutSubTileViews.gridSubTile,
+                            rewriter.getArrayAttr({})},
+        /*bounds=*/ArrayRef<int64_t>{1, 1, 1, 1, elementsInThreadBuffer},
+        /*strides=*/ArrayRef<int64_t>{1, 1, 1, 1, 1},
+        /*useIndexDiffs=*/true, /*forceUnroll=*/true);
+    {
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.setInsertionPointToStart(loop.getBody());
+
+      Block::BlockArgListType gemm0OutCoords = loop.getLowerCoords(0);
+      Block::BlockArgListType upperCoords = loop.getLowerCoords(1);
+      auto isLargerThanM = rewriter.create<arith::CmpIOp>(
+          loc, arith::CmpIPredicate::sge, gemm0OutCoords[1], prePadGemmMVal);
+      auto isLargerThanN = rewriter.create<arith::CmpIOp>(
```

Comment:
You shouldn't need this? You could just query the `validity` argument of the transforming_for, which'll return `false` in the event of padding?

Heck, you could extend `global_load` with a `pad_value : T`, which you'd use in the case of invalidity (though then you'd want to not use buffer ops, since those only support 0 and sometimes 1 as padding values)

Reply:
To rephrase, could you *create* the maps that would have been used if the intermediate tensor *were* written to global memory, and then use the validity bit those generate to do this `-inf` change?